nodestream/schema/migrations/auto_change_detector.py

```diff
@@ -0,0 +1,580 @@
+        properties_attr: str = "non_key_properties",
+    ):
+        def detect_changes_between_types(from_type_name, to_type_name):
+            from_type_props = getattr(from_population[from_type_name], properties_attr)
+            to_type_props = getattr(to_population[to_type_name], properties_attr)
+            # To start, we assume that all properties missing from one side or
+            # the other were created or deleted.
+            deleted = from_type_props - to_type_props
+            added = to_type_props - from_type_props
+            renamed = set()
+
+            # We don't have any smart way to ascertain what is probably
+            # a rename, so we are just going to have to ask about every
+            # combination of properties
+            for added_property in added:
+                for deleted_property in deleted:
+                    if self.input.ask_property_renamed(
+                        to_type_name, deleted_property, added_property
+                    ):
+                        renamed.add((to_type_name, deleted_property, added_property))
+
+            # For renamed properties, we can remove them from the
+            # added and removed sets and add it to to the final renamed set.
+            for item in renamed:
+                _, deleted_property, added_property = item
+                deleted.remove(deleted_property)
+                added.remove(added_property)
+                renamed_property_location.add(item)
+
+            # We can add the type and the property to the final set once
+            # we have removed renamed properties.
+            for property in added:
+                added_property_loction.add((to_type_name, property))
+            for property in deleted:
+                deleted_property_location.add((to_type_name, property))
+
+        for from_type_name, to_type_name in self.compare_types_pairwise(
+            from_population,
+            to_population,
+            deleted_types_location,
+            new_types_location,
+            renamed_types_location,
+        ):
+            detect_changes_between_types(from_type_name, to_type_name)
+
+    def detect_type_changes(
+        self,
+        from_population: Dict[str, GraphObjectSchema],
+        to_population: Dict[str, GraphObjectSchema],
+        deleted_types_location: Set[str],
+        renamed_types_location: Set[Tuple[str, str]],
+        new_types_location: Set[str],
+    ):
+        # To start, we assume that all types missing from one side or
+        # the other were created or deleted.
+        current_types = set(to_population)
+        previous_types = set(from_population)
+        deleted_types_location.update(previous_types - current_types)
+        new_types_location.update(current_types - previous_types)
+
+        # After that, check for types that appear to have been renamed
+        # where the type names are different but the keys are the same.
+        for new_type_name in new_types_location:
```

Comment:
What happens if new_type is renamed, but one of the keys is also changed?

Reply:
Good question... 

As it is today, this would be detected as a drop of the original node type and a create of the rename of the node type. 

```
Generated migration 20240206124201.
The migration contains 4 schema changes.
  - Drop node type Person
  - Create node type Human
  - Drop additional index for node type Person on field last_ingested_at
  - Add additional index for node type Human on field last_ingested_at
```

Reply:
I guess the change could simply remove the check if the keys match...

```diff
-   keys_match = new_type_def.has_matching_keys(prev_type_def)
-   if keys_match and self.input.ask_type_renamed(
+   if self.input.ask_type_renamed(
          prev_type_name, new_type_name
     ):
```

That does produce a more correct detection:

```
❯ poetry run nodestream migrations make
Did you rename Person to Human? (yes/no) [no] yes
Did you rename first_name to first on Human? (yes/no) [no] yes
Generated migration 20240206124704.
The migration contains 2 schema changes.
  - Rename node type Person to Human
  - Rename key part first_name to first on node type Human
Migration written to migrations/20240206124704.yaml
Run `nodestream migrations run` to apply the migration.
```